qiskit/synthesis/permutation/permutation_utils.py

```diff
@@ -0,0 +1,41 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Utility functions for handling permutations."""
+
+
+def _get_ordered_swap(permutation_in):
+    """Sorts the input permutation by iterating through the permutation list
+    and putting each element to its correct position via a SWAP (if it's not
+    at the correct position already). If ``n`` is the length of the input
+    permutation, this requires at most ``n`` SWAPs.
+
+    More precisely, if the input permutation is a cycle of length ``m``,
+    then this creates a quantum circuit with ``m-1`` SWAPs (and of depth ``m-1``);
+    if the input  permutation consists of several disjoint cycles, then each cycle
+    is essentially treated independently.
+    """
+    permutation = list(permutation_in[:])
+    swap_list = []
+    for i, val in enumerate(permutation):
+        if val != i:
+            j = permutation.index(i)
```

Comment:
I would avoid using the `index()` call here, it basically is an `O(n)`  lookup on each iteration making this loop scale quadratically. If you need to get the index of `i` from permutation, I would build a map outside the loop that can do this look up in constant time. That way this function is `O(2n)` instead of `O(n**2)`.

Reply:
This is a good point. For the record, this function existed before, I have moved it to `permutation_utils.py` without any modifications. But I agree that it would be nice to make it more efficient. Hmm, it might be just a tiny bit trickier than what you have suggested, as the permutation (and hence also `index()`) changes throughout the loop, but it's clear that a linear algorithm should be possible.

Reply:
And that was very easy, done in [15f1228](https://github.com/Qiskit/qiskit-terra/pull/9082/commits/15f1228d6791aaa0223c8fd040a3658c61b7185a).